DRAFT - 22.0.0.4-BETA BLOG (#2080)
OpenLiberty/blogs: posts/2022-03-17-beta-release-post.adoc

---
layout: post
title: "TITLE"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ryan-storey
author_github: https://github.com/ryan-storey
seo-title: MicroProfile GraphQL 2.0, Java 18 support, and distributed security caches on Open Liberty 22.0.0.4-beta - OpenLiberty.io

> seo-description: Open Liberty 22.0.0.4 offers GraphQL 2.0, incorporating Jakarta EE 9.1 dependencies, as well as Java 18 support with many new exciting features and changes. Also included in this release is distributed security caches, allowing multiple Liberty servers to share caches via a JCache provider.

Reviewer: ```suggestion
seo-description: Open Liberty 22.0.0.4-beta offers GraphQL 2.0, incorporating Jakarta EE 9.1 dependencies, as well as Java 18 support with many new exciting features and changes. Also included in this release is distributed security caches, allowing multiple Liberty servers to share caches via a JCache provider.
```

> blog_description: Open Liberty 22.0.0.4 offers GraphQL 2.0, incorporating Jakarta EE 9.1 dependencies, as well as Java 18 support with many new exciting features and changes. Also included in this release is distributed security caches, allowing multiple Liberty servers to share caches via a JCache provider.

Reviewer: ```suggestion
blog_description: Open Liberty 22.0.0.4-beta offers GraphQL 2.0, incorporating Jakarta EE 9.1 dependencies, as well as Java 18 support with many new exciting features and changes. Also included in this release is distributed security caches, allowing multiple Liberty servers to share caches via a JCache provider.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= MicroProfile GraphQL 2.0, Java 18 support, and distributed security caches on Open Liberty 22.0.0.4-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 22.0.0.4 offers GraphQL 2.0, incorporating Jakarta EE 9.1 dependencies, as well as Java 18 support with many new exciting features and changes. Also included in this release is distributed security caches, allowing multiple Liberty servers to share caches via a JCache provider.

Reviewer: ```suggestion
Open Liberty 22.0.0.4-beta offers GraphQL 2.0, incorporating Jakarta EE 9.1 dependencies, as well as Java 18 support with many new exciting features and changes. Also included in this release is distributed security caches, allowing multiple Liberty servers to share caches via a JCache provider.
```

The link:{url-about}[Open Liberty] 22.0.0.4-beta includes the following beta features (along with all GA features):

* <<graphql, MicroProfile GraphQL 2.0>>
* <<java18, Java 18 Support>>
* <<security_caches, Distributed Security Caches>>

[#graphql]
== MicroProfile GraphQL 2.0

MicroProfile GraphQL 2.0 incorporates Jakarta EE 9.1 dependencies.  This allows developers to continue using the same functionality provided by MP GraphQL 1.0 but with the updated Jakarta components such as CDI 3.0, Jakarta REST 3.0, JSON-B 2.0, etc. Functionally, version 2.0 behaves the same as 1.0 because it uses the same version of the underlying implementation (link:https://github.com/smallrye/smallrye-graphql[SmallRye GraphQL]).  If you'd like to learn more about MP GraphQL, see our previous link:https://openliberty.io/blog/2020/06/10/microprofile-graphql-open-liberty.html[blog post about MP GraphQL 1.0] and our link:https://openliberty.io/blog/2020/08/28/graphql-apis-open-liberty-20009.html[blog post on how to use the GraphiQL UI and third-party Client APIs].

To enable the new beta features in your app, add them to your `server.xml`:

[source, xml]
----
<featureManager>

>   <feature>mpGraphQL-1.0</feature>

Reviewer: ```suggestion
  <feature>mpGraphQL-2.0</feature>
```

</featureManager>
----

[#java18]
== Java 18 Support

Java 18 is coming soon, and with it the following features and changes:

* link:https://openjdk.java.net/jeps/400[UTF-8 by Default]
* link:https://openjdk.java.net/jeps/408[Simple Web Server]
* link:https://openjdk.java.net/jeps/413[Code Snippets in Java API Documentation]
* link:https://openjdk.java.net/jeps/416[Reimplement Core Reflection with Method Handles]
* link:https://openjdk.java.net/jeps/417[Vector API (Third Incubator)]
* link:https://openjdk.java.net/jeps/418Internet-Address Resolution SPI]
* link:https://openjdk.java.net/jeps/419[Foreign Function & Memory API (Second Incubator)]
* link:https://openjdk.java.net/jeps/420[Pattern Matching for switch (Second Preview)]
* link:https://openjdk.java.net/jeps/421[Deprecate Finalization for Removal]

By taking advantage of trying out the new changes in Java 18 now, you get more time to review your applications, microservices and runtime environments and be a step ahead when it becomes generally available.

> To give this a try today, link:https://jdk.java.net/18/[download the early access release of Java 18], download and install the link:https://openliberty.io/downloads/#runtime_betas[22.0.0.4-beta] version of Open Liberty, edit your Liberty server's link:https://openliberty.io/docs/22.0.0.4/reference/config/server-configuration-overview.html#server-env[server.env file] to point `JAVA_HOME` to your Java 18 installation and start testing!

Reviewer: ```suggestion
To give this a try today, link:https://jdk.java.net/18/[download the early access release of Java 18], <<run, install  22.0.0.4-beta>> version of Open Liberty, edit your Liberty server's link:https://openliberty.io/docs/latest/reference/config/server-configuration-overview.html#server-env[server.env file] to point `JAVA_HOME` to your Java 18 installation and start testing!
```

For more information on Java 18, please visit the Java 18 link:https://jdk.java.net/18/release-notes[release notes page], link:https://download.java.net/java/early_access/jdk18/docs/api/[API Javadoc page] or link:https://jdk.java.net/18/[download page].
For more information on Open Liberty, please visit our link:https://openliberty.io/docs[documentation page].

> NOTE: As we work toward full Java 18 support ourselves, please excuse any of our functionality that might not be 100% ready yet.

Reviewer: ```suggestion
If you find anything amiss, <<feedback, let us know>>.
```

[#security_caches]
== Distributed Security Caches

Distributed security cache support has been introduced so that multiple Liberty servers can share caches via a `JCache` provider.
Prior to this feature, the authentication (subject) and logged out cookie caches were restricted to be local and in-memory. Multiple servers were unable to benefit from their peers' caches and each server started with a cold cache.

As part of this feature both caches can now both be stored in a distributed `JCache`. This can improve performance and failure recovery, reduce the load on backend user registries and improve the security posture of the server.

> NOTE: When using the new distribute caching functionality in this beta release, the feature `distributedSecurityCache-1.0` needs to be included in the list of features in the `server.xml` file. This beta-only feature that is used to expose the `JCache` APIs to the 3rd party `JCache` provider. When the distributed caching functionality is GA'd, this feature will no longer be available.

Reviewer: ```suggestion
NOTE: When using the new distribute caching functionality in this beta release, the feature `distributedSecurityCache-1.0` needs to be included in the list of features in the `server.xml` file. This beta-only feature exposes the `JCache` APIs to the 3rd party `JCache` provider. When the distributed caching functionality is released as part of a GA, it will be exposed in a different manner then via the `distributedSecurityCache-1.0` feature, which will no longer be available.
```

Reviewer: @dmuelle will likely be able to further clarify/clean that up.  We could even just drop that last sentence.

Because the creation of a subject might affect performance, Liberty provides an authentication cache to store a subject after an authentication of a user is successful. The authentication cache now can be distributed using a 3rd party JCache provider. To configure the distributed authentication cache, use the following server.xml configuration:

[source, xml]
----
    <featureManager>
        <feature>appSecurity-3.0</feature>
        <feature>distributedSecurityCache-1.0</feature>
    </featureManager>

    <!-- 
        The 3rd-party JCache provider library that Liberty will use to manage and connect to the cache.
     -->
    <library id="JCacheProviderLib">
        <fileset dir="${shared.resource.dir}" includes="jcacheprovider.jar" />
    </library>

    <!-- 
        Configure the JCache cache instance.
     -->
    <cache id="AuthCache" name="AuthCache">
        <cacheManager uri="uri://someuri">
            <properties prop1="value1" prop2="value2" />

            <cachingProvider libraryRef="JCacheProviderLib" />
        </cacheManager>
    </cache>

    <!-- 
        Configure the authentication cache.
     -->
    <authCache cacheRef="AuthCache" />
----

> If your Liberty environment injects custom principals or credentials into your subject (for instance in a custom LoginModule or Trust Association Interceptor (`TAI`)), in order to store them in the distributed authentication cache, they will need to be `Serializable`. Additionally, it is important that the shared library that contains those classes is made available to the caching provider and any other configurations that need access to those classes. If the same shared library is not used for each, `ClassCastExceptions` could be encountered when working with the classes retrieved from the distributed cache.

Reviewer: ```suggestion
If your Liberty environment injects custom principals or credentials into your subject (for instance in a custom `LoginModule` or Trust Association Interceptor (`TAI`)), in order to store them in the distributed authentication cache, they will need to be `Serializable`. Additionally, it is important that the shared library that contains those classes is made available to the caching provider and any other configurations that need access to those classes. If the same shared library is not used for each, `ClassCastExceptions` could be encountered when working with the classes retrieved from the distributed cache.
```

    are stored in the subject.
 -->
<library id="CustomLib">
    <fileset dir="${shared.resource.dir}" includes="customlibrary.jar" />
</library>

<!-- 
    Take notice that the 'libraryRef' attribute has both library references.
 -->
<cache ... >
    <cacheManager ... >
        <cachingProvider libraryRef="JCacheProviderLib,CustomLib" />
    </cacheManager>
</cache>

<!--
    Some sample JAAS custom login module configuration. The custom login module
    in this example would inject custom credentials or principals into the subject.

    Note that the 'libraryRef' in the 'jaasLoginModule' needs to be set to the same
    library referenced from the caching provider.
 -->
<jaasLoginContextEntry id="system.WEB_INBOUND"
    name="system.WEB_INBOUND"
    loginModuleRef="custom, hashtable, userNameAndPassword, certificate, token" />

<jaasLoginModule id="custom"
    className="org.acme.CustomLoginModule"
    controlFlag="REQUIRED" libraryRef="CustomLib" />

<!-- 
    Any applications that will be accessing classes from the Subject also need
    to use the same library reference.
 -->
<application ...>
    <classloader commonLibraryRef="CustomLib" />
</application>
----

> A few points to consider when configuring a JCache cache for use with the authentication cache.

Reviewer: ```suggestion
A few points to consider when configuring a JCache for use with the authentication cache.
```

* The distributed authentication cache is comprised of keys and values of type `Object`
to match the behavior of the local authentication cache, set a least recently used eviction (`LRU`) policy with a maximum entry count of 25000 and an entry TTL of 600 seconds. Note that with distributed caches, partitioning of the cache can lead to an actual capacity below the configured value.
* If your `JCache` provider supports it, configure a client-side cache to reduce transactions to the distributed cache. If the client-side cache supports storing the entries as deserialized objects, this can further improve performance.
* Support in the beta is currently limited to `LTPA` and `JWT`. Single sign-on methods such as `SPNEGO`, `Oauth`, `OIDC` and `SAML` (etc) are not yet supported
* Subjects in the distributed cache should be treated as you would treat other security-sensitive information, such as usernames and passwords. Configure your `JCache` provider to secure the data while it is in motion and at rest. This should include encryption and access control.

*Configuring a Distributed Logged-Out Cookie Cache*

> The logged-out cookie cache stores `LTPA` and `JWT` cookies that have been logged-out. The logged-out cookie cache can now be distributed using a 3rd party `JCache` provider ensuring that logged out cookies are enforced across multiple servers. To configure the distributed logged-out cookie cache, use the following `server.xml` configuration:

Reviewer: ```suggestion
*Configuring a Distributed Logged-Out Cookie Cache*

The logged-out cookie cache stores `LTPA` and `JWT` cookies that have been logged-out. The logged-out cookie cache can now be distributed using a 3rd party `JCache` provider ensuring that logged out cookies are enforced across multiple servers. To configure the distributed logged-out cookie cache, use the following `server.xml` configuration:
```

[source, xml]
----
    <featureManager>
        <feature>appSecurity-3.0</feature>
        <feature>distributedSecurityCache-1.0</feature>
    </featureManager>

    <!-- 
        The 3rd-party JCache provider library that Liberty will use to manage and connect to the cache.
     -->
    <library id="JCacheProviderLib">
        <fileset dir="${shared.resource.dir}" includes="jcacheprovider.jar" />
    </library>

    <!-- 
        Configure the JCache instances.
     -->
    <cache id="LoggedOutCookieCache" name="LoggedOutCookieCache">
        <cacheManager uri="uri://someuri">
            <properties prop1="value1" prop2="value2" />

            <cachingProvider libraryRef="JCacheProviderLib" />
        </cacheManager>
    </cache>

    <!-- 
        Configure the authentication cache to use the JCache. 
     -->
    <webAppSecurity loggedoutCookieCacheRef="LoggedOutCookieCache" />
----

A few points to consider when configuring a JCache cache for use with the logged-out cookie cache.

> * The distributed logged-out cookie cache is comprised of keys and values of type Object

Reviewer: ```suggestion
* The distributed logged-out cookie cache is comprised of keys and values of type `Object`
```

        <feature>distributedSecurityCache-1.0</feature>
        <feature>sessionCache-1.0</feature>
    </featureManager>

    <!-- 
        The 3rd-party JCache provider library that Liberty will use to manage and connect to the cache.
     -->
    <library id="JCacheProviderLib">
        <fileset dir="${shared.resource.dir}" includes="jcacheprovider.jar" />
    </library>

    <!-- 
        Configure the JCache cache manager.
     -->
    <cacheManager id="CacheManager" uri="uri://someuri">
        <properties prop1="value1" prop2="value2" />

        <cachingProvider libraryRef="JCacheProviderLib" />
    </cacheManager>

    <!-- 
        Configure the JCache cache instances.
     -->
    <cache id="AuthCache" name="AuthCache" cacheManagerRef="CacheManager" />
    <cache id="LoggedOutCookieCache" name="LoggedOutCookieCache" cacheManagerRef="CacheManager" />

    <!-- 
        Configured the authentication cache, logged-out cookie cache and HTTP session cache.
     -->
    <authCache cacheRef="AuthCache" />
    <webAppSecurity loggedoutCookieCacheRef="LoggedOutCookieCache" ... />
    <httpSessionCache cacheManagerRef="CacheManager" ... />
----

> To find out more, check out link:https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-configuring-authentication-cache-in[Configuring the authentication cache in Liberty], and link:https://www.ibm.com/docs/en/was-liberty/base?topic=manually-configuring-liberty-session-persistence-jcache[Configuring Liberty session persistence with JCache].

Reviewer: ```suggestion
To find out more, check out the link:https://openliberty.io/docs/latest/reference/config/authentication.html[authentication] and link:https://openliberty.io/docs/latest/reference/config/authCache.html[authCache] elements enabled by the link:https://www.openliberty.io/docs/latest/reference/feature/appSecurity.html[appSecurity] feature, and the link:https://www.openliberty.io/docs/latest/reference/feature/sessionCache.html#_examples[JCache Session Persistence examples].
```

> === Try it now 

Reviewer: ```suggestion
[#run]
=== Try it now 
```

> To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 17, Java SE 11, or Java SE 8.

Reviewer: ```suggestion
To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 18, Java SE 11, or Java SE 8.
```
